src/messaging/ExchangeContext.cpp

```diff
@@ -689,5 +698,21 @@ void ExchangeContext::ExchangeSessionHolder::GrabExpiredSession(const SessionHan
     GrabUnchecked(session);
 }
 
+bool ExchangeContext::IsPeerLikelyActiveHint()
+{
+    // If we are not the initiator, it means the peer sent the first message.
+    // This means our peer is active if we are sending a message to them.
```

Comment:
Is this true?  Let's say the peer sends the first message.  We send the response.  The peer gets the response, sends the ack, closes its exchange, goes to sleep.

The ack gets lost.  We are going to retry with MRP.  What timeout should we use?

I realize this is aiming to be a heuristic, and it looks like it's optimizing for the case when our response got lost as opposed to the ack getting lost, right?  But we should be really clear in our documentation what things are actually known to be true vs what things we assume to be true vs what things we just don't know but are making a choicer.

But again: this is not special to initiators.  Let's say _we_ are the initiator.  We sent message A, peer responded with message B, now we are sending message C.  The fact that we are sending a response to B is probably a good indication that the peer is still active waiting for that response, just like the logic for the "peer is initiator" case.

So the real test here is more like "if we are sending a response" (i.e. we have received a message on this exchange before), not "if we are not the initiator".

Reply:
Yes, just basing off the IsInitiator() is incorrect. Then the message received potentially has to be the first message in the Exchange. If the UnsolicitedMessageHandler is processing a message, then we can probably make this conclusion, which we are probably making by assuming the peer to be active at the point of sending the response anyway.

Reply:
> So the real test here is more like "if we are sending a response" (i.e. we have received a message on this exchange before), not "if we are not the initiator".

This ends adds up to the same thing since the function is only called when we are sending a message. If we are sending a message which we know here is not an ack and we are not the initiator, we know for a fact that the peer is going to be active.

The initiator flag is for the entire lifetime of the exchange context and does change at any point in time.


> Let's say the peer sends the first message. We send the response. The peer gets the response, sends the ack, closes its exchange, goes to sleep.
The ack gets lost. We are going to retry with MRP. What timeout should we use?

In this case, the initiator would retry at the `IdleRestransTimeout` since we can't know if the peer is active or not.

Reply:
> and we are not the initiator

My point is: it does not matter whether we are the initiator.  What matters is whether this is the very first message being sent on this exchange or not, no?

Reply:
As @bzbarsky-apple also stated above, this algorithm is "optimizing for the case when our response got lost as opposed to the ack getting lost,". Assuming the ack getting lost, then we may send all the retries while the peer is idle (because the peer has sent the ack and it went to idle mode). 

If the response can be piggyback with the ack of the initial message, wouldn't be wiser to consider the peer in idle mode afterwards?

ICD -> Client: Send a Message
Client -> ICD: Send Rsp + piggyback ACK => If this msg is lost then the client will retry.
ICD -> Client: ACK for the Rsp => If this message is lost then the ICD went idle and we should retry later.

Reply:
@doru91 i don't think i understand you comment...

If we receive the response the message we sent, we assume the peer is idle unless we are told otherwise.

> Client -> ICD: Send B (as response) + piggyback ACK for A => client will use IdleRestransTimeout   for retries. If this msg is lost then the ICD will retransmit A.

This is incorrect. Since the client isn't the initiator, it knows the peer (ICD in this case) is active and will use the `ActiveRestransTimeout`.

Reply:
resolving comment since it is no longer applicable

Reply:
> This is incorrect. Since the client isn't the initiator, it knows the peer (ICD in this case) is active and will use the ActiveRestransTimeout.

It is not incorrect, it was the proposal: instead of using `ActiveRestransTimeout` to use `IdleRestransTimeout `, but _only_ for the cases where the response can be piggyback with the ACK for the initial message. 

Let me rephrase it. Current behavior using the PR changes:

1. ICD -> Client: Send A
2. Client -> ICD: Send B (as response) + piggyback ACK for A => client will use `ActiveRestransTimeout` for retries. If this message is lost then the assumption that the ICD is still active it's the right decision. However, if the next message (3) is lost then all the retries may be send while the ICD is idle.
3. ICD -> Client: ACK for B => Considering that ICD has an `ActiveModeThreshold ` value of 0, ICD will go idle immediately after sending the ACK.

Wouldn't be safer to:
1. ICD -> Client: Send A
2. Client -> ICD: Send B (as response) + piggyback ACK for A => client will use `IdleRestransTimeout` for retries. If the initial message is lost, then the ICD will retransmit A fast (because its ACK was lost). If the next message (3) is lost then a retry will be sent at the next active time of the ICD.
3. ICD -> Client: ACK for B => Considering that ICD has an `ActiveModeThreshold ` value of 0, ICD will go idle immediately after sending the ACK.

Reply:
> the consequence of doing that is the risk of the ICD staying in ActiveMode when it could have gotten back to IdleMode

` if `IdleRestransTimeout ` is used for the retransmits of message 2, it doesn't necessarily mean that the retransmits will be done at that rate, it only protects against flooding the icd while in idle mode if message 3 got lost: that's because retransmits of message 2 will be automatically triggered by retransmits of message 1 (which is not receiving its ACK). However, this is assuming the lost rate is identical in both directions. I will resolve the conversation for now as this solution is dependent on the lost rate.

Reply:
I think the underlying problem is how MRP functions. I think MRP requires some work to make it adequate for all use-case.